nntrainer/graph/graph_core.h

```diff
@@ -205,6 +205,26 @@ class GraphCore {
    */
   void addLossToSorted();
 
+  void addInputNode(std::shared_ptr<GraphNode> &node) {
```

Comment:
I wonder if the graph_core needs this function? can graph_core extract this information by itself? I recommend to modify the modify graph core interface for graph core to interpret itself given the nodes, which are input nodes and which are output nodes.

Reply:
you might have to introduce a function (getOutputConnections() like getInputConnections()) in GraphNode.

Reply:
So you means instread of addInputNode and addOutputNode which is called in realizeGraph by network_graph make a function in graph_core something like findinputOutputNode and call that function after running the realizeGraph. And getOutputConnection will be needed when making that function am I right?

Reply:
yes, you are right.

Reply:
I've done what you said but check replaceOutputNode function that this also needs to removed. Actually we can also delete this function by realize input output node after addLossLayer. But in this case in the we didn't have a information which layer is a output node so we need to check all the layer to find output layer in addLossLayer

Reply:
Yes, that will be good. `addLossLayer()` can once iterate over all the nodes to find output layers - it's fine IMO.
It's an optimization that can be done in the next PR as this PR is already big enough.